File: leetcode/problems/language-specific/go/11.container-with-most-water.go
/*
 * @lc app=leetcode id=11 lang=golang
 *
 * [11] Container With Most Water
 */

// @lc code=start
func maxArea(height []int) int {
	res, left, right := 0, 0, len(height)-1
	for left < right {
		// * Area of rectangle = length * width
		area := findMin(height[left], height[right]) * (right - left)
		res = findMax(res, area)
		// * Move the shorter height pointer
		if height[left] < height[right] {
			left++
		} else {
			right--
		}
	}

	return res
}

func findMax(a, b int) int {
	if a >= b {
		return a
	}

	return b
}

func findMin(a, b int) int {
	if a <= b {
		return a
	}

	return b
}

// @lc code=end

File: leetcode/problems/sessions/2025-10-25_2025-??-??/11.container-with-most-water.py
#
# @lc app=leetcode id=11 lang=python3
#
# [11] Container With Most Water
#

# @lc code=start

# * Two Pointers Solution | O(n) Time | O(1) Space
# * n -> The length of height array


class Solution:
    def maxArea(self, height: List[int]) -> int:
        left, right = 0, len(height) - 1
        max_area = 0
        while left < right:
            # * Max area = height * width
            cur_area = min(height[left], height[right]) * (right - left)
            max_area = max(max_area, cur_area)
            if height[left] < height[right]:
                left += 1
            else:
                right -= 1

        return max_area


# @lc code=end
